### ISSUE TITLE
Add new peer organization workflow fails

### ISSUE THREAD

**Issue by gabrieleCacciola** [APGGroeiFabriek/PIVT#24]
I followed the instructions up to the following command:

`helm template peer-org-flow/ -f samples/scaled-raft-tls/configtx.yaml -f samples/scaled-raft-tls/crypto-config.yaml -f samples/scaled-raft-tls/network.yaml -f samples/scaled-raft-tls/hostAliases.yaml | argo submit - --watch`

and I get the following output:

> Name:                hlf-peer-orgs-m92jk
Namespace:           blockchain
ServiceAccount:      default
Status:              Failed
Message:             child 'hlf-peer-orgs-m92jk-1388713836' failed
Created:             Wed Nov 27 21:37:24 +0000 (4 seconds ago)
Started:             Wed Nov 27 21:37:24 +0000 (4 seconds ago)
Finished:            Wed Nov 27 21:37:28 +0000 (now)
Duration:            4 seconds

STEP                                                                                              PODNAME                         DURATION  MESSAGE
 ✖ hlf-peer-orgs-m92jk (peer-orgs)                                                                                                          child 'hlf-peer-orgs-m92jk-1388713836' failed
 └---✖ consortiums (consortiums)                                                                                                            child 'hlf-peer-orgs-m92jk-1012712537' failed
     └---✖ SecondConsortium (consortium--SecondConsortium)                                                                                  child 'hlf-peer-orgs-m92jk-3318890625' failed
         └---✖ consortium-org--Karga (consortium-org--SecondConsortium--Karga)                                                              Invalid 'when' expression '!{{steps.check-org-in-consortium.outputs.result}}': Invalid token: '!{{'
             ├---✔ check-org-in-consortium(0) (check-org-in-consortium--SecondConsortium--Karga)  hlf-peer-orgs-m92jk-3341541617  3s
             └---⚠ add-org-to-consortium (add-org-to-consortium--SecondConsortium--Karga)                                                   Invalid 'when' expression '!{{steps.check-org-in-consortium.outputs.result}}': Invalid token: '!{{'

Any hint?

PS: Is it possible to add OrdererOrgs using the same set of steps?

**Comment by raftAtGit** [comment]
check this one https://github.com/APGGroeiFabriek/PIVT/issues/19, possibly same issue.
i.e. you didnt create `rolebinding` forArgo

adding an OrdererOrg is somewhat similar but you cannot use this flow for that. 
I plan to implement a similar flow for adding OrdererOrgs but no timeline yet.

**Comment by gabrieleCacciola** [comment]
Hi,

That doesn't seem to be the issue unfortunately:

> UK-GCACCIOLA-MBP:fabric-kube gabrielecacciola$ kubectl create rolebinding default-admin --clusterrole=admin --serviceaccount=argo:default
Error from server (AlreadyExists): rolebindings.rbac.authorization.k8s.io "default-admin" already exists

**Comment by raftAtGit** [comment]
can u please check the logs of the failing Argo side car container
`kubectl logs hlf-peer-orgs-m92jk-1388713836 -c wait`

**Comment by gabrieleCacciola** [comment]
`UK-GCACCIOLA-MBP:fabric-kube gabrielecacciola$ kubectl logs hlf-peer-orgs-m92jk-3341541617 -c wait -n blockchain
time="2019-11-27T21:37:27Z" level=info msg="Creating a docker executor"
time="2019-11-27T21:37:27Z" level=info msg="Executor (version: v2.4.2, build_date: 2019-10-21T18:39:09Z) initialized (pod: blockchain/hlf-peer-orgs-m92jk-3341541617) with template:\n{\"name\":\"check-org-in-consortium--SecondConsortium--Karga\",\"arguments\":{},\"inputs\":{},\"outputs\":{},\"metadata\":{},\"script\":{\"name\":\"\",\"image\":\"hyperledger/fabric-tools:1.4.1\",\"command\":[\"bash\"],\"env\":[{\"name\":\"FABRIC_LOGGING_SPEC\",\"value\":\"info\"},{\"name\":\"CORE_PEER_LOCALMSPID\",\"value\":\"GroeifabriekMSP\"},{\"name\":\"CORE_PEER_TLS_CERT_FILE\",\"value\":\"/etc/hyperledger/fabric/tls/server.crt\"},{\"name\":\"CORE_PEER_TLS_KEY_FILE\",\"value\":\"/etc/hyperledger/fabric/tls/server.key\"},{\"name\":\"CORE_PEER_TLS_ROOTCERT_FILE\",\"value\":\"/etc/hyperledger/fabric/tls/ca.crt\"}],\"resources\":{},\"volumeMounts\":[{\"name\":\"orderer-tlsca\",\"mountPath\":\"/hlf_config/orderer-tlsca\"},{\"name\":\"orderer-groeifabriek-orderer0-tls\",\"mountPath\":\"/etc/hyperledger/fabric/tls/\"},{\"name\":\"orderer-groeifabriek-admin-msp\",\"mountPath\":\"/etc/hyperledger/fabric/msp/\"},{\"name\":\"hlf-scripts\",\"mountPath\":\"/hlf-scripts\"}],\"source\":\"mkdir -p /work \\u0026\\u0026 \\n\\npeer channel fetch config /work/config_block.pb -c testchainid -o orderer0.groeifabriek.nl:7050 \\\\\\n        --tls --cafile /hlf_config/orderer-tlsca/tlscacert.pem \\\\\\n\\u0026\\u0026 \\u003e\\u00262 echo '-- fetched config block from orderer system channel testchainid' \\u0026\\u0026\\n\\nconfigtxlator proto_decode --input /work/config_block.pb --type common.Block \\\\\\n      | jq .data.data[0].payload.data.config \\u003e /work/config.json \\u0026\\u0026\\necho \\u003e\\u00262 '-- trimmed envelope from config block and converted to json' \\u0026\\u0026\\n\\nconsortium=$(jq  '.channel_group.groups.Consortiums.groups.\\\"SecondConsortium\\\"' /work/config.json) \\u0026\\u0026\\nif test \\\"$consortium\\\" == \\\"null\\\"; then\\n  echo \\u003e\\u00262 '-- failed to parse consortium SecondConsortium from config block, does it exist?'\\n  exit 1\\nfi\\n\\norg=$(echo $consortium | jq '.groups.KargaMSP') \\u0026\\u0026\\nif test \\\"$org\\\" == \\\"null\\\"; then\\n  echo \\u003e\\u00262 '-- peer organization Karga does not exist in the consortium SecondConsortium'\\n  echo false\\nelse\\n  echo \\u003e\\u00262 '-- peer organization Karga exists in the consortium SecondConsortium'\\n  echo true\\nfi\\n\"},\"retryStrategy\":{\"limit\":4}}"
time="2019-11-27T21:37:27Z" level=info msg="Waiting on main container"
time="2019-11-27T21:37:27Z" level=error msg="executor error: Failed to establish pod watch: unknown (get pods)\ngithub.com/argoproj/argo/errors.Wrap\n\t/go/src/github.com/argoproj/argo/errors/errors.go:88\ngithub.com/argoproj/argo/errors.InternalWrapErrorf\n\t/go/src/github.com/argoproj/argo/errors/errors.go:78\ngithub.com/argoproj/argo/workflow/executor.(*WorkflowExecutor).waitMainContainerStart\n\t/go/src/github.com/argoproj/argo/workflow/executor/executor.go:916\ngithub.com/argoproj/argo/workflow/executor.(*WorkflowExecutor).Wait\n\t/go/src/github.com/argoproj/argo/workflow/executor/executor.go:880\ngithub.com/argoproj/argo/cmd/argoexec/commands.waitContainer\n\t/go/src/github.com/argoproj/argo/cmd/argoexec/commands/wait.go:40\ngithub.com/argoproj/argo/cmd/argoexec/commands.NewWaitCommand.func1\n\t/go/src/github.com/argoproj/argo/cmd/argoexec/commands/wait.go:16\ngithub.com/spf13/cobra.(*Command).execute\n\t/go/src/github.com/spf13/cobra/command.go:766\ngithub.com/spf13/cobra.(*Command).ExecuteC\n\t/go/src/github.com/spf13/cobra/command.go:852\ngithub.com/spf13/cobra.(*Command).Execute\n\t/go/src/github.com/spf13/cobra/command.go:800\nmain.main\n\t/go/src/github.com/argoproj/argo/cmd/argoexec/main.go:17\nruntime.main\n\t/usr/local/go/src/runtime/proc.go:201\nruntime.goexit\n\t/usr/local/go/src/runtime/asm_amd64.s:1333"
time="2019-11-27T21:37:27Z" level=info msg="No output parameters"
time="2019-11-27T21:37:27Z" level=info msg="No output artifacts"
time="2019-11-27T21:37:27Z" level=info msg="No Script output reference in workflow. Capturing script output ignored"
time="2019-11-27T21:37:27Z" level=info msg="Killing sidecars"
time="2019-11-27T21:37:27Z" level=warning msg="Failed to get pod 'hlf-peer-orgs-m92jk-3341541617': pods \"hlf-peer-orgs-m92jk-3341541617\" is forbidden: User \"system:serviceaccount:blockchain:default\" cannot get resource \"pods\" in API group \"\" in the namespace \"blockchain\""
time="2019-11-27T21:37:27Z" level=error msg="Failed to kill sidecars: pods \"hlf-peer-orgs-m92jk-3341541617\" is forbidden: User \"system:serviceaccount:blockchain:default\" cannot get resource \"pods\" in API group \"\" in the namespace \"blockchain\""
time="2019-11-27T21:37:27Z" level=info msg="Alloc=4981 TotalAlloc=10118 Sys=70078 NumGC=3 Goroutines=6"`

**Comment by raftAtGit** [comment]
looks like the same issue. 

you are running in a different namespace `blockchain`, not `default`. I guess you need to create `rolebinding` in that namespace

**Comment by gabrieleCacciola** [comment]
Still failing unfortunately...

`UK-GCACCIOLA-MBP:fabric-kube gabrielecacciola$ kubectl -n blockchain logs hlf-peer-orgs-rg68f-931367991 -c init
time="2019-11-27T23:20:53Z" level=info msg="Creating a docker executor"
time="2019-11-27T23:20:53Z" level=info msg="Executor (version: v2.4.2, build_date: 2019-10-21T18:39:09Z) initialized (pod: blockchain/hlf-peer-orgs-rg68f-931367991) with template:\n{\"name\":\"check-org-in-consortium--SecondConsortium--Karga\",\"arguments\":{},\"inputs\":{},\"outputs\":{},\"metadata\":{},\"script\":{\"name\":\"\",\"image\":\"hyperledger/fabric-tools:1.4.1\",\"command\":[\"bash\"],\"env\":[{\"name\":\"FABRIC_LOGGING_SPEC\",\"value\":\"info\"},{\"name\":\"CORE_PEER_LOCALMSPID\",\"value\":\"GroeifabriekMSP\"},{\"name\":\"CORE_PEER_TLS_CERT_FILE\",\"value\":\"/etc/hyperledger/fabric/tls/server.crt\"},{\"name\":\"CORE_PEER_TLS_KEY_FILE\",\"value\":\"/etc/hyperledger/fabric/tls/server.key\"},{\"name\":\"CORE_PEER_TLS_ROOTCERT_FILE\",\"value\":\"/etc/hyperledger/fabric/tls/ca.crt\"}],\"resources\":{},\"volumeMounts\":[{\"name\":\"orderer-tlsca\",\"mountPath\":\"/hlf_config/orderer-tlsca\"},{\"name\":\"orderer-groeifabriek-orderer0-tls\",\"mountPath\":\"/etc/hyperledger/fabric/tls/\"},{\"name\":\"orderer-groeifabriek-admin-msp\",\"mountPath\":\"/etc/hyperledger/fabric/msp/\"},{\"name\":\"hlf-scripts\",\"mountPath\":\"/hlf-scripts\"}],\"source\":\"mkdir -p /work \\u0026\\u0026 \\n\\npeer channel fetch config /work/config_block.pb -c testchainid -o orderer0.groeifabriek.nl:7050 \\\\\\n        --tls --cafile /hlf_config/orderer-tlsca/tlscacert.pem \\\\\\n\\u0026\\u0026 \\u003e\\u00262 echo '-- fetched config block from orderer system channel testchainid' \\u0026\\u0026\\n\\nconfigtxlator proto_decode --input /work/config_block.pb --type common.Block \\\\\\n      | jq .data.data[0].payload.data.config \\u003e /work/config.json \\u0026\\u0026\\necho \\u003e\\u00262 '-- trimmed envelope from config block and converted to json' \\u0026\\u0026\\n\\nconsortium=$(jq  '.channel_group.groups.Consortiums.groups.\\\"SecondConsortium\\\"' /work/config.json) \\u0026\\u0026\\nif test \\\"$consortium\\\" == \\\"null\\\"; then\\n  echo \\u003e\\u00262 '-- failed to parse consortium SecondConsortium from config block, does it exist?'\\n  exit 1\\nfi\\n\\norg=$(echo $consortium | jq '.groups.KargaMSP') \\u0026\\u0026\\nif test \\\"$org\\\" == \\\"null\\\"; then\\n  echo \\u003e\\u00262 '-- peer organization Karga does not exist in the consortium SecondConsortium'\\n  echo false\\nelse\\n  echo \\u003e\\u00262 '-- peer organization Karga exists in the consortium SecondConsortium'\\n  echo true\\nfi\\n\"},\"retryStrategy\":{\"limit\":4}}"
time="2019-11-27T23:20:53Z" level=info msg="Loading script source to /argo/staging/script"
time="2019-11-27T23:20:53Z" level=info msg="Start loading input artifacts..."
time="2019-11-27T23:20:53Z" level=info msg="Alloc=4713 TotalAlloc=9849 Sys=70078 NumGC=3 Goroutines=4"`

**Comment by raftAtGit** [comment]
what is failing now exactly?

in your previous logs there was this one, which was caused by missing `rolebinding`, but in your last logs it's gone
```
level=error msg="Failed to kill sidecars: pods \"hlf-peer-orgs-m92jk-3341541617\" is forbidden: User \"system:serviceaccount:blockchain:default\" cannot get resource \"pods\" in API group \"\" in the namespace \"blockchain\""
```

**Comment by gabrieleCacciola** [comment]
I created the `rolebinding` successfully but still having issues with the workflow, same output.

```UK-GCACCIOLA-MBP:fabric-kube gabrielecacciola$ kubectl create rolebinding default-admin --clusterrole=admin --serviceaccount=argo:default -n blockchain
rolebinding.rbac.authorization.k8s.io/default-admin created```

```Name:                hlf-peer-orgs-mmhxm
Namespace:           blockchain
ServiceAccount:      default
Status:              Failed
Message:             child 'hlf-peer-orgs-mmhxm-3676404078' failed
Created:             Wed Nov 27 23:29:02 +0000 (4 seconds ago)
Started:             Wed Nov 27 23:29:02 +0000 (4 seconds ago)
Finished:            Wed Nov 27 23:29:05 +0000 (1 second ago)
Duration:            3 seconds

STEP                                                                                              PODNAME                         DURATION  MESSAGE
 ✖ hlf-peer-orgs-mmhxm (peer-orgs)                                                                                                          child 'hlf-peer-orgs-mmhxm-3676404078' failed
 └---✖ consortiums (consortiums)                                                                                                            child 'hlf-peer-orgs-mmhxm-2861734475' failed
     └---✖ SecondConsortium (consortium--SecondConsortium)                                                                                  child 'hlf-peer-orgs-mmhxm-847061335' failed
         └---✖ consortium-org--Karga (consortium-org--SecondConsortium--Karga)                                                              Invalid 'when' expression '!{{steps.check-org-in-consortium.outputs.result}}': Invalid token: '!{{'
             ├---✔ check-org-in-consortium(0) (check-org-in-consortium--SecondConsortium--Karga)  hlf-peer-orgs-mmhxm-1879091891  3s
             └---⚠ add-org-to-consortium (add-org-to-consortium--SecondConsortium--Karga)                                                   Invalid 'when' expression '!{{steps.check-org-in-consortium.outputs.result}}': Invalid token: '!{{'```

**Comment by raftAtGit** [comment]
I guess this is not correct:
```
kubectl create rolebinding default-admin --clusterrole=admin --serviceaccount=argo:default -n blockchain
```
it should be:
```
kubectl create rolebinding default-admin --clusterrole=admin --serviceaccount=blockchain:default -n blockchain
```

**Comment by gabrieleCacciola** [comment]
Thanks, seems like we moved a bit further:

```Name:                hlf-peer-orgs-h6zhk
Namespace:           blockchain
ServiceAccount:      default
Status:              Failed
Message:             child 'hlf-peer-orgs-h6zhk-1015336098' failed
Created:             Wed Nov 27 23:35:07 +0000 (10 seconds ago)
Started:             Wed Nov 27 23:35:07 +0000 (10 seconds ago)
Finished:            Wed Nov 27 23:35:17 +0000 (now)
Duration:            10 seconds

STEP                                                                                                                           PODNAME                         DURATION  MESSAGE
 ✖ hlf-peer-orgs-h6zhk (peer-orgs)                                                                                                                                       child 'hlf-peer-orgs-h6zhk-1015336098' failed
 └---✖ consortiums (consortiums)                                                                                                                                         child 'hlf-peer-orgs-h6zhk-642175471' failed
     └---✖ SecondConsortium (consortium--SecondConsortium)                                                                                                               child 'hlf-peer-orgs-h6zhk-2203377875' failed
         ├---✔ consortium-org--Karga (consortium-org--SecondConsortium--Karga)
         |   ├---✔ check-org-in-consortium(0) (check-org-in-consortium--SecondConsortium--Karga)                               hlf-peer-orgs-h6zhk-28949623    3s
         |   └---○ add-org-to-consortium (add-org-to-consortium--SecondConsortium--Karga)                                                                                when '!true' evaluated false
         └---✖ consortium-org--Valhalla (consortium-org--SecondConsortium--Valhalla)                                                                                     child 'hlf-peer-orgs-h6zhk-475202304' failed
             ├---✔ check-org-in-consortium(0) (check-org-in-consortium--SecondConsortium--Valhalla)                            hlf-peer-orgs-h6zhk-1797578135  4s
             └---✖ add-org-to-consortium (add-org-to-consortium--SecondConsortium--Valhalla)                                                                             child 'hlf-peer-orgs-h6zhk[0].consortiums[0].SecondConsortium[1].consortium-org--Valhalla[1].add-org-to-consortium[0].create-system-channel-config-update' errored
                 ├---● create-system-channel-config-update (create-system-channel-config-update--SecondConsortium--Valhalla)
                 └-⚠ create-system-channel-config-update(0) (create-system-channel-config-update--SecondConsortium--Valhalla)  hlf-peer-orgs-h6zhk-2897078544  0s        controller is not configured with a default archive location```

**Comment by raftAtGit** [comment]
well, as mentioned in requirements, you need to setup an artifactory for Argo for backup/restore and peer-org flows.

easiest is to install Minio with Helm.

I would suggest first setup an artifactory for Argo, make sure an Argo sample with artifact works fine then come back to peer-org-flow

**Comment by gabrieleCacciola** [comment]
Damn, sorry I misread that piece, I thought it was just for backup/restore and not for the peer-org flow.

I will proceed with your suggestions and eventually (hopefully not!) open another ticket.

**Comment by gabrieleCacciola** [comment]
PS: thanks for your support!

**Comment by raftAtGit** [comment]
welcome :)

artifactory is needed when some file is needed to be copied from one step to another. in backup/restore obviously we need to copy backed up content, for peer-org-flow, we need to copy the signed config update to next steps.
